src/basic/errno-util.h

```diff
@@ -6,6 +6,14 @@
 
 #include "macro.h"
 
+/* strerror(3) says that glibc uses a maximum length of 1024 bytes. */
+#define ERRNO_BUF_LEN 1024
+
+/* Note: the lifetime of the compound literal is the immediately surrounding block,
+ * see C11 §6.5.2.5, and
+ * https://stackoverflow.com/questions/34880638/compound-literal-lifetime-and-if-blocks */
+#define STRERROR(errnum) strerror_r(abs(errnum), (char[ERRNO_BUF_LEN]){}, ERRNO_BUF_LEN)
```

Comment:
maybe clarify that this only works using the GNU version of the helper.

maybe even add a compile-check somehow that insists that _GNU_SOURCE is set, because only then you'll get the GNU version. Or alternatively maybe even better would be an `assert_cc()` check on `__builtin_types_compatible_p()` on `typeof(strerror_r)` and the expected type that returns a const string

Reply:
also, please add a comment clarifying what this returns if I pass garbage to this, i.e. error INT_MAX or so

Reply:
If the other variant was used, then we'd immediately get a long list of type errors. We explicitly turn printf format checks on, so we get this for free.

Reply:
well, i was more interested in codifying this for the reader to note in the header file than the actual errors. i.e. the implicit assumptions here should be written down explicitly here. could be a comment, could be such an assert.

Reply:
please add a comment here at the very least, this deserves a note for the reader, since we are on purpose preferring the GNU version over the POSIX version.

(I wonder how musl folks will like htis)

Reply:
I added a comment and a test.

Reply:
Right at this spot:
```C
/* strerror(3) says that glibc uses a maximum length of 1024 bytes. */
#define ERRNO_BUF_LEN 1024

/* Note: the lifetime of the compound literal is the immediately surrounding block,
 * see C11 §6.5.2.5, and
 * https://stackoverflow.com/questions/34880638/compound-literal-lifetime-and-if-blocks
 *
 * Note that we use the GNU variant of strerror_r() here. */
#define STRERROR(errnum) strerror_r(abs(errnum), (char[ERRNO_BUF_LEN]){}, ERRNO_BUF_LEN)

/* A helper to print an error message or message for functions that return 0 on EOF.
 * Note that we can't use ({ … }) to define a temporary variable, so errnum is
 * evaluated twice. */
#define STRERROR_OR_EOF(errnum, message) (errnum) != 0 ? STRERROR(errnum) : (message)
```